components/experimental/src/units/ratio.rs

```diff
@@ -10,45 +10,77 @@ use core::{
+
+use super::provider::{Base, ConversionInfoULE, SiPrefix, Sign};
 
-// TODO: implement AsULE for IcuRatio and use it in Data module.
 // TODO: add test cases for IcuRatio.
 /// A ratio type that uses `BigInt` as the underlying type.
 #[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
 pub struct IcuRatio(Ratio<BigInt>);
 
 #[derive(Debug)]
 pub enum IcuRatioError {
+    /// Represents an error when parsing a `BigInt` from a string.
+    /// For example, the string "1/23" is invalid because it contains a '/' character.
+    ///              the string "3A" is invalid because it contains a non-numeric character.
     BigIntParseError(num_bigint::ParseBigIntError),
+
+    /// Represents an error when parsing a ratio from a string.
+    /// For example, the string "1/0/2" is invalid because it contains more than one '/' character.
     InvalidRatioString,
 }
 
 impl IcuRatio {
-    pub fn new(numerator: BigInt, denominator: BigInt) -> Self {
+    /// Creates a new `IcuRatio` from the given numerator and denominator.
+    pub(crate) fn from_big_ints(numerator: BigInt, denominator: BigInt) -> Self {
         Self(Ratio::new(numerator, denominator))
     }
 
-    pub fn numerator(&self) -> &BigInt {
-        self.0.numer()
-    }
-
-    pub fn numerator_str(&self) -> String {
-        self.0.numer().to_string()
-    }
-
-    pub fn denominator(&self) -> &BigInt {
-        self.0.denom()
-    }
-
-    pub fn denominator_str(&self) -> String {
-        self.0.denom().to_string()
+    /// Extracts the factor ratio from `ConversionInfoULE`.
+    pub(crate) fn factor_from_conversion_info(conversion_info: &ConversionInfoULE) -> Self {
+        let sign: num_bigint::Sign = Sign::from_unaligned(conversion_info.factor_sign).into();
+        Self::from(Ratio::<BigInt>::new(
+            BigInt::from_bytes_le(sign, conversion_info.factor_num().as_ule_slice()),
+            BigInt::from_bytes_le(
+                num_bigint::Sign::Plus,
+                conversion_info.factor_den().as_ule_slice(),
+            ),
+        ))
+    }
+
+    /// Extracts the offset ratio from `ConversionInfoULE`.
+    pub(crate) fn offset_from_conversion_info(conversion_info: &ConversionInfoULE) -> Self {
+        let sign: num_bigint::Sign = Sign::from_unaligned(conversion_info.offset_sign).into();
+        Self::from(Ratio::<BigInt>::new(
+            BigInt::from_bytes_le(sign, conversion_info.offset_num().as_ule_slice()),
+            BigInt::from_bytes_le(
+                num_bigint::Sign::Plus,
+                conversion_info.offset_den().as_ule_slice(),
+            ),
+        ))
+    }
+
+    /// Applys the given SI prefix to the ratio.
+    /// For example: if the ratio is 1/3 and the SI prefix is "kilo" (10^3),
+    ///              the result will be 1000/3.
+    /// Another example: if the ratio is 1/3 and the SI prefix is "kibi" ("2^10"),
+    ///                  the result will be 1024/3.
+    pub(crate) fn apply_si_prefix(&mut self, si_prefix: &SiPrefix) {
```

Comment:
nit: I'd prefer if this was `impl MulAssign` between `IcuRatio` and `&SiPrefix`, because this is inherently a multiplication.

Alternatively, you could have a `SiPrefix::as_icu_ratio` that does the match, and multiply with that.

Reply:
done